File: lib/empydom.js
// empydom.js
// This is the JavaScript side of the empydom DOM bridge

(function() {

  // the charBuffer is a buffer until a newline character from Python
  var charBuffer = ""
    // a blob is an object with properties 'val' and 'isPrimitive'
    var blobs = {};
  blobs[0] = {};
  blobs[0].val = window;
  blobs[0].isPrimitive = false;
  var nextID = 1;

  var capitalize = function(lowercase) {
    var str = "" + lowercase;
    return str.charAt(0).toUpperCase() + str.slice(1);
  };

  var convertToPythonObject = function (obj) {
    if (typeof(obj) === "boolean"){
      if (obj === true) return "True";
      else return "False";
    }
    if (typeof(obj) === "number") {
      return obj;
    }
    else { //typeof(obj) === "string"
      return JSON.stringify(obj);
    }
  };

  var returnBlobToPython = function (localBlobID, pythonBlobID) {
    Python.eval("pyjs.blobs[" + pythonBlobID + "] = (" + localBlobID + ", " + convertToPythonObject(blobs[localBlobID].isPrimitive) + ")");
  };

  var throwInPython = function(errorStack) {
    Python.eval('pyjs.error = """' + errorStack + '"""')
  };

  var setPrimitiveInPython = function(value) {
  };

  var objIsPrimitive = function(obj){
    return (typeof(obj) === "string" ||
        typeof(obj) === "boolean" ||
        typeof(obj) === "number");
  };

  var createBlob = function (val){
    var currID = nextID;
    blobs[currID] = {};
    blobs[currID].val = val;
    if (objIsPrimitive(blobs[currID].val)){
      blobs[currID].isPrimitive = true;
    }
    else {
      blobs[currID].isPrimitive = false;
    }
    nextID += 1;
    return currID;
  };

  var makeAlias = function(object, name) {
    var fn = object ? object[name] : null;
    if (typeof fn == 'undefined')
      return function() {};
    return function() {
      return fn.apply(object, arguments);
    }
  };

  var swapQuotes = function(strWithQuotes) {
    return strWithQuotes.replace(/'/g, '\u2190').replace(/"/g, "'").replace(/\u2190/g, '"');
  };

  var convertJSONtoObj = function(json) {

    json = swapQuotes(json);
    //if JSON.parse succeeds assume it's an object
    try {
      var obj = JSON.parse(json);
      json = obj;
    }
    catch(e) { }
    return json;
  };

  var templateToFunc = function(arg) {
    var CANARY_LEN = 10;

    if (/^==%%==%%==/.test(arg)) {
      arg = eval("(" + arg.substr(CANARY_LEN) + ")");
    }

    return arg;
  };

  var countSpacesBeforeWordChar = function(str) {
    var spaces = 0;
    for (var i = str.length-1; i >= 0; i--) {
      if (str.charAt(i) == '\n') {
        spaces = 0;
      }
      else if (str.charAt(i) == ' ') {
        spaces += 1;
      }
      else {
        return spaces;
      }
    }
  };

  var fixSpacing = function(code, spaces) {
    var lines = code.split("\n");

    lines[0] = lines[0] + "\n";
    for (var i = lines.length-1; i >= 1; i--) {
      lines[i] = lines[i].substr(spaces) + "\n";
    }
    return lines.join("");
  };

  var module = {
    'blobs': blobs,
    //if Python is not initialized than use default chrToFunc
    //if chrToFunc !== undefined then we assume you want to initialize it in here
    'executePythonScripts': function(chrToFunc) {
      if (chrToFunc === undefined && !Python.isInitialized) {
        chrToFunc = function(chr) {
          if (chr !== null) {
            realChar = String.fromCharCode(chr);
            charBuffer += realChar;
            if (realChar == '\n') {
              console.log(charBuffer);
              charBuffer = "";
            }
          }
        };
      }

      if (chrToFunc !== undefined) {
        Python.initialize(null, chrToFunc);
      }

      //from https://developer.mozilla.org/En/Using_web_workers
      Array.prototype.forEach.call(
          document.querySelectorAll("script[type=\"text\/python\"]"),
          function(scriptTag) {
            var spacesBefore = countSpacesBeforeWordChar(scriptTag.innerHTML);
            var pycode = scriptTag.innerHTML.trim();
            pycode = fixSpacing(pycode, spacesBefore);
            Python.eval(pycode);
          });

    },
    'delBlob': function(sourceBlobID) {
      //TODO: Make this function similar to Python's `del` we only want 
      //      to decrement the reference count, not delete the property

      //try {
      //delete blobs[sourceBlobID];
      //}
      //catch(e) {
      //throwInPython(e.toString());
      //}
    },
    'explodeArgs': function(argList) {
      var explodedArgs = argList[0];
      for (var i = 1; i < argList.length; i++) {
        explodedArgs += "," + argList[i];
      }
      if (explodedArgs === undefined) {
        explodedArgs = "";
      }
      return explodedArgs;
    },
    //Objects passed from Python to JavaScript can be taken to be immutable
    //They are pass by value
    'createBlobFromJSON': function(jsonStr, otherBlobID) {
      try {
        var currID = createBlob(JSON.parse(jsonStr));
        returnBlobToPython(currID, otherBlobID);
      }
      catch (e) {
        throwInPython(e.stack);
      }
    },
    'createBlobFromBlobProperty': function(sourceBlobID, property, otherBlobID){
      try {
        var currID = createBlob(blobs[sourceBlobID].val[property]);
        returnBlobToPython(currID, otherBlobID);
      }
      catch(e) {
        throwInPython(e.stack);
      }
    },
    'callBlobFunction': function(parentBlobID, localBlobID, otherBlobID, args){

      try {
        for (var i = 0; i < args.length; i++) {
          args[i] = convertJSONtoObj(args[i]);
          args[i] = templateToFunc(args[i]);
        }

        //TODO: make this fix less hacky (window.document(10) won't throw an error now)
        if (blobs[localBlobID].val.apply !== undefined) {
          var currID = createBlob(blobs[localBlobID].val.apply(blobs[parentBlobID].val, args));
          returnBlobToPython(currID, otherBlobID);
        }
      }
      catch(e) {
        throwInPython(e.stack);
      }
    },
    'getBlobPrimitive': function(localBlobID) {
      try {
        var primitive = blobs[localBlobID].val;
        Python.eval("pyjs.primitive = " + convertToPythonObject(primitive));
      }
      catch(e) {
        throwInPython(e.stack);
      }
    },
    'setBlobPropertyToValue': function(localBlobID, property, value) {
      try {
        blobs[localBlobID].val[property] = value;
      }
      catch(e) {
        throwInPython(e.stack);
      }
    }
  };

  return module;

})();

File: lib/empydom.py
import bifrost
 
class JSWrapper:
  def __init__(self, pyjs, jsID, parentJSID):
    self.createLocal__("pyjs__")
    self.createLocal__("jsID__")
    self.createLocal__("parentJSID__")
    self.pyjs__ = pyjs
    self.jsID__ = jsID
    self.parentJSID__ = parentJSID

  def createLocal__(self, attr):
    self.__dict__[attr] = None

  def __setattr__(self, attr, value):
    if attr in self.__dict__:
      self.__dict__[attr] = value
    else :
      self.pyjs__.setJSBlobPrimitive(self.jsID__, attr, value)

  def __getattr__(self, attr):
    newID = self.pyjs__.getJSBlobProperty(self.jsID__, attr)
    return self.getWrappedBlob__(newID)

  def __call__(self, *args):
    newID = self.pyjs__.callJSBlob(self.parentJSID__, self.jsID__, args)
    return self.getWrappedBlob__(newID)

  def __del__(self):
    self.pyjs__.delBlob(self.jsID__)

  def getWrappedBlob__(self, id):
    isPrimitive = self.pyjs__.blobs[id][1]
    if isPrimitive:
      primitive = self.pyjs__.getJSBlobPrimitive(id)
      #TODO: delete the reference
      return primitive
    else :
      return JSWrapper(self.pyjs__, id, self.jsID__)


class Empydom:
  def __init__(self):
    self.windowId = 0
    self.nextId = 1
    self.blobs = {0: (self.windowId, False)}
    self.primitive = None
    self.error = None
    self.funcs = []

  def jserror(self):
    err = self.error
    self.nextId -= 1 #throw out that blob index
    self.error = None #reset error
    raise Exception(err)

  def call(self, jsstring):
    bifrost.run(jsstring)

  def delBlob(self, jsID):
    self.call("window.empydom.delBlob(" + str(jsID) + ")")
    del self.blobs[jsID]

  #rewrite, messy hack at 5am
  def callJSBlob(self, parentjsid, jsid, args):
    jsID = self.blobs[jsid][0]
    jsArgs = []
    for arg in args:
      jsArgs.append(self.convertToJSObject(arg,True))
    currentId = self.nextId
    self.call("window.empydom.callBlobFunction(" + str(parentjsid) +", " + str(jsid) + ", " + str(currentId) + ", " + str(jsArgs) + ")") 
    self.nextId += 1
    if (self.error is not None):
      self.jserror()
    return currentId

  def getJSBlobProperty(self, jsID, prop): 
    currentId = self.nextId
    self.call("window.empydom.createBlobFromBlobProperty(" + str(jsID) + ", \"" + str(prop) + "\", " + str(currentId) + ")") 
    self.nextId += 1
    if (self.error is not None):
      self.jserror()
    return currentId

  def getJSBlobPrimitive(self, pythonID):
    jsID = self.blobs[pythonID][0]
    self.call("window.empydom.getBlobPrimitive(" + str(jsID) + ")")
    if (self.error is not None):
      self.jserror()
    return self.primitive

  def setJSBlobPrimitive(self, pythonID, prop, value):
    jsID = self.blobs[pythonID][0]
    self.call("window.empydom.setBlobPropertyToValue(" + str(jsID) + ", \"" + str(prop) + "\", " + self.convertToJSObject(value,False) + ")")
    if (self.error is not None):
      self.jserror()

  def wrapFunction(self, func):
    self.funcs.append(func)
    index = len(self.funcs)-1
    return """==%%==%%==function() {var argString = window.empydom.explodeArgs(arguments); Python.eval( "pyjs.funcs[""" + str(index) + """](" + argString + ")");}"""

  #If the object is in a list than even if str,
  #it needs to be same due to str() reprs
  def convertToJSObject(self, pythonObj, isInList):
    if (type(pythonObj) == str and isInList == False):
      return "\"" + pythonObj + "\""
    elif (type(pythonObj) == type(lambda x: x)):
      return self.wrapFunction(pythonObj)
    return str(pythonObj)

pyjs = Empydom()
window = JSWrapper(pyjs, pyjs.windowId, pyjs.windowId)

